lib/linalg.dx

```diff
@@ -65,14 +110,14 @@ def chol {n} (x:n=>n=>Float) : LowerTriMat n Float =
     mat = lower_tri_mat buf
     for i:n. for j':(..i).
       j = inject n j'
-      row  = for k:(..<j). get $ mat i (cast k)
-      row' = for k:(..<j). get $ mat j (cast k)
-      a = x.i.j - vdot row row'
-      if (ordinal i) == (ordinal j)
+      row_i = for k:(..<j'). get $ mat i (inject (..i) k)
+      row_j = for k:(..<j'). get $ mat j (unsafe_cast k)
```

Comment:
Is this unsafe cast covered by a subset mapping instance?
```
instance {m n} [Subset m n] Subset (..<m) (..<n)
  inject' = \x. unsafe_from_ordinal _ $ ordinal x
  project' = \x. if ordinal x < get_size m then
    then Just $ unsafe_from_ordinal _ $ ordinal x
    else Nothing
```

Reply:
OK, on second thought, that can't be right.  What I was trying to spell was more along the lines of
```
instance {m n} [Ix m, Ix n, Subset m n] {i:m} {j:n} Subset (..<i) (..<j)
```
but that's simply not true.

What's really going on is that when we write
```
j = inject n j'
```
we are leaving on the floor a proof object that allows us to type
```
j':(..j)
```
If we had that typing for `j'`, then we could just `inject (..j) k` using the subset instance for `Subset (..<n) (..n)`, but we don't have such a type, and I don't know how to adjust the type system to give it to us.

That means we can fall back to a helper that captures the use we wanted to put that proof to:
```
def dual_inject {m n} [Ix m, Ix n, Subset m n] (i:m) (j:(..<i)) : LowerTriIxExc n =
  i' = inject n i
  j' = from_just $ project _ $ inject n $ inject m j
  MkLowerTriIxExc i' j'
```

I notice that this requires another law for `inject`, namely that it's compatible with the ordering implied by `Ix`.  Maybe we should capture that law in another class?

I'm also none too pleased with how non-reusable these implicit proof objects are turning out to be.  For instance, if we had two indices we were trying to carry along with the original injection into `n`, I think we would need another helper.

Reply:
thanks for this, it worked!  I agree that this all seems a little baroque.  But I like this workflow of unsafe -> safe but unreadable -> ??? -> safe and readable